src/selectors_vm/stack.rs

```diff
@@ -52,13 +52,96 @@ impl ChildCounter {
     }
 }
 
+struct CounterItem {
+    /// The counter at this index
+    pub counter: ChildCounter,
+    /// The index of this counter in the stack
+    pub index: usize,
+}
+
+struct CounterList {
+    items: Vec<CounterItem>,
+    // we always have at least one item, an empty list shouldn't exist in the map
+    current: CounterItem,
+}
+
+#[derive(Default)]
+/// A more efficient counter that only requires one owned local name to track counters across multiple stack frames
+pub struct TypedChildCounterMap(HashMap<LocalName<'static>, CounterList>);
+
+impl TypedChildCounterMap {
+    /// Adds a seen child to the map. The index is the level of the item
+    pub fn add_child(&mut self, name: &LocalName, index: usize) {
+        let hash = {
+            let mut hasher = self.0.hasher().build_hasher();
+            name.hash(&mut hasher);
+            hasher.finish()
+        };
+        let entry = self.0.raw_entry_mut().from_hash(hash, |n| name == n);
+        match entry {
+            RawEntryMut::Vacant(vacant) => {
+                let mut counter = ChildCounter::default();
+                counter.inc();
+                vacant.insert_hashed_nocheck(
+                    hash,
+                    name.clone().into_owned(), // the hash won't change just because we've got ownership
+                    CounterList {
```

Comment:
the majority of the code in this branch can be refactored into `CounterList::new()`